File: src/Monet/MOTimer.m
#import <Monet/MOTimer.h>

#import <SDL/SDL.h>

struct MOTimerData
{
	SDL_TimerID	timerID;

	uint32_t    interval;
	double      duration;

	void        *userInfo;
};

Uint32 timerCallback(Uint32 interval, void *param)
{
	MOTimer *timer = (MOTimer *)param;

	SDL_Event event;
	event.type = SDL_USEREVENT;
	event.user.data1 = timer;
	SDL_PushEvent(&event);

	return [timer duration];
}

@implementation MOTimer

- (id)initWithDuration: (double)aDuration userInfo: (void *)aUserInfo
{
	if ((self = [super init]))
	{
		timerData = calloc(1, sizeof(struct MOTimerData));

		timerData->duration = aDuration;
		timerData->userInfo = aUserInfo;
	}

	return self;
}

- (void)dealloc
{
	[self stop];
	free(timerData);

	[super dealloc];
}

#pragma mark -

- (double)duration
{
	return timerData->duration;
}

- (void *)userInfo
{
	return timerData->userInfo;
}

#pragma mark -

- (void)start
{
	timerData->timerID = SDL_AddTimer(
		(uint32_t)(timerData->duration * 1000.0),
		&timerCallback,
		self);
}

- (void)stop
{
	SDL_RemoveTimer(timerData->timerID);
}

@end

File: src/Monet/MOTimer.c
#include <Monet/MOTimer.h>

#include <cobject/cobject.h>
#include <SDL/SDL.h>

struct _MOTimer
{
	COGuts      *guts;

	SDL_TimerID	timerID;

	uint32_t    interval;
	double      duration;

	void        *userInfo;
};

void _MOTimerDestroy(void *aTimer);

Uint32 _MOTimerFired(uint32_t aInterval, void *aParam)
{
	MOTimer *timer = (MOTimer *)aParam;

	SDL_Event event;
	event.type = SDL_USEREVENT;
	event.user.data1 = timer;
	SDL_PushEvent(&event);

	return MOTimerGetDuration(timer);
}

MOTimer *MOTimerCreate(double aDuration, void *aUserInfo)
{
	MOTimer *timer = malloc(sizeof (MOTimer));
	COInitialize(timer);
	COSetDestructor(timer, &_MOTimerDestroy);

	timer->duration = aDuration;
	timer->userInfo = aUserInfo;

	return timer;
}

void _MOTimerDestroy(void *aTimer)
{
	MOTimer *timer = (MOTimer *)timer;

	MOTimerStop(timer);
}

double MOTimerGetDuration(MOTimer *aTimer)
{
	return aTimer->duration;
}

void *MOTimerGetUserInfo(MOTimer *aTimer)
{
	return aTimer->userInfo;
}

void MOTimerStart(MOTimer *aTimer)
{
	aTimer->timerID = SDL_AddTimer(
		(uint32_t)(aTimer->duration * 1000.0l),
		&_MOTimerFired,
		aTimer);
}

void MOTimerStop(MOTimer *aTimer)
{
	SDL_RemoveTimer(aTimer->timerID);
}
